### PR TITLE
Avoid compile error in old GCC (CentOS6)

### PR DESCRIPTION
**Pull request by okuta**
I got following error on CentOS6.
```
cupy/core/_scalar.cpp:964: error: member '__pyx_t_float_complex __pyx_t_4cupy_4core_7_scalar_Scalar::complex64_' with constructor not allowed in union
cupy/core/_scalar.cpp:965: error: member '__pyx_t_double_complex __pyx_t_4cupy_4core_7_scalar_Scalar::complex128_' with constructor not allowed in union
```

### CODE REVIEW

File: cupy/core/_scalar.pxd [hunk 1]
```diff
@@ -1,35 +1,11 @@
...
-    int32_t int32_
-    int64_t int64_
-    uint8_t uint8_
-    uint16_t uint16_
-    uint32_t uint32_
-    uint64_t uint64_
-    float float32_
-    double float64_
-    float complex complex64_
-    double complex complex128_
```
Comment (kmaehashi) [inline cupy/core/_scalar.pxd]: Note for later reference: These two `complex` lines are causes of this issue; unrestricted unions are implemented in GCC 4.6 whereas CentOS 6 only provides GCC 4.4.

File: cupy/core/_scalar.pyx [hunk 2]
```diff
@@ -114,54 +145,54 @@ cdef class CScalar(CPointer):
...
+                s.uint32_ = val_u
             elif size == 8:
-                self.val.uint64_ = val_u
+                s.uint64_ = val_u
             else:
                 assert False
         elif kind == 'f':
             if size == 2:
-                self.val.uint16_ = <uint16_t>internal.to_float16(<float>val)
+                s.uint16_ = _float16.to_float16(<float>val)
```
Comment (kmaehashi) [inline cupy/core/_scalar.pyx]: `s/_float16/internal/`

File: cupy/core/_scalar.pyx [hunk 3]
```diff
@@ -37,7 +62,7 @@ cdef dict _dtype_kind_size_dict = {}
 
 cdef _setup_type_dict():
     cdef char k
-    for i in _dtype.all_type_chars:
+    for i in _dtype_module.all_type_chars:
```
Comment (kmaehashi) [inline cupy/core/_scalar.pyx]: Is it better to change `all_types_chars` to `cdef` to avoid `_dtype_module`? (correct me if I'm wrong)

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
This line expects the existance of `CScalar.val`.
https://github.com/cupy/cupy/blob/b63795a0b4d1e63bc2894ddfbe4988e89c7743b7/cupy/core/reduction.pxi#L160-L161

**Comment by okuta** [comment]
Thanks! I fixed. @kmaehashi

**Comment by kmaehashi** [comment]
Thanks!
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM!

**Pull request merged**